src/pgwire/src/protocol.rs

```diff
@@ -308,16 +314,28 @@ where
                 state = match state {
                     State::Ready => self.advance_ready().await?,
                     State::Drain => self.advance_drain().await?,
+                    State::TimedOut(err) => self.advance_timed_out(err).await?,
                     State::Done => return Ok(()),
-                }
+                };
+                self.adapter_client()
```

Comment:
A thought: timeouts should live in SessionClient. This will make this feature work over http, and only require setting up timeouts when interacting with coord instead of on every pgwire message.

Reply:
Good idea, I'll play around with this. It's worth noting that as far as I can tell the HTTP API doesn't allow interactive transactions because it closes the session after every query. So this specific timeout isn't possible on the HTTP API, but future timeouts, like `statement_timeout` might be.

Reply:
If a pgwire client runs a `SELECT` but then just doesn't read rows off the network, would that prevent this idle detection from working? Should we care? How do we handle large result sets or slow client networks? Can you research what postgres defines (in the code) as idle? I do think it's reasonable for idle to mean "time between the client sending any message to the server", even if part of that time is taken up by the client waiting for the server. If users hit that a lot (due to, say, very large responses), they can increase the idle timeout.

Reply:
The two places that Postgres starts the idle timer are below:
https://github.com/postgres/postgres/blob/2fe3bdbd691a5d11626308e7d660440be6c210c8/src/backend/tcop/postgres.c#L4394-L4396
https://github.com/postgres/postgres/blob/2fe3bdbd691a5d11626308e7d660440be6c210c8/src/backend/tcop/postgres.c#L4407-L4409

These are called if the client is in an aborted transaction or active transaction, right before they block and wait for the client to send a command. This is right below where the timer is started:
https://github.com/postgres/postgres/blob/2fe3bdbd691a5d11626308e7d660440be6c210c8/src/backend/tcop/postgres.c#L4480-L4483

I think this nicely maches are start time which begins right before awaiting the client connection for a frontend message.

Postgres stops the timer once it receives a command from the client:
https://github.com/postgres/postgres/blob/2fe3bdbd691a5d11626308e7d660440be6c210c8/src/backend/tcop/postgres.c#L4485-L4497

This also matches are stop time, which is triggered right after we've awaited the client connection for a frontend message.

So I would say that Postgres defines idle as the time the backend spends waiting for a command from the frontend.


> If a pgwire client runs a SELECT but then just doesn't read rows off the network, would that prevent this idle detection from working? Should we care? How do we handle large result sets or slow client networks?

I'm assuming that Postgres sends query results synchronously, so therefore it won't wait for the next command until the result is fully sent, so therefore large results and slow networks will not cause the idle timeout to trigger. If the `psql` process is just letting the query results sit in it's TCP buffer without sending back ACKs, then it also won't trigger this timeout.

However, Postgres has two additional timeout that cover this situation:
- The external `statement_timout` (https://www.postgresql.org/docs/current/runtime-config-client.html) which aborts a statement that takes too long to fully execute.
- The internal `CLIENT_CONNECTION_CHECK_TIMEOUT` https://github.com/postgres/postgres/blob/fbfe6910eca0d6a61cbcdd27cdd5a8d9de6477fa/src/include/utils/timeout.h#L36
which I don't fully understand, but seems to timeout non-responsive connections.

Reply:
This is a great summary. Seems like if we want to exactly copy postgres this must live in our pgwire crate. We could make a choice to not mirror them exactly. There is a bit of appeal to moving this to SessionClient which resets the timer each time it sends a command to the coordinator. That has a nice property of "it doesn't matter what the reason is (slow network, large result set, or idle client), clients must send commands to the thing that is keeping open the txn or they get timedout", and it'd work with the http client which I think also might be susceptible to large result problems.

New problem: how will these implementations affect long running SUBSCRIBE statements in a txn (say, because a user is using `DECLARE` and `FETCH` in a loop)? `FETCH` by default waits for some results. What if no results come in during the timeout period? In the `SessionClient` impl we would time them out because we didn't see any user request (which would be wrong). In the `pgwire` case we wouldn't start the timeout timer while waiting for the `FETCH`, which is correct. Ok, so maybe this means we should absolutely not do SessionClient?

Reply:
> There is a bit of appeal to moving this to SessionClient which resets the timer each time it sends a command to the coordinator.

If I understand this correctly, that would mean that we would include the time it takes to process the query in the idle time. I'm not a huge fan of this tbh, because we could end up with situations where we terminate a connection and return an error saying the client was idle, when in reality it just took us a long time to process their query. I think this scenario would be better handled by the `statement_timeout`. In my mind, all time either falls into idle time or statement time. So with those two timeouts we'd cover everything.

Having said that, if it's possible to implement the idle timeout in the `SessionClient`, without including processing time, I'd much prefer that. Even if it means differing from the Postgres implementation. I was just unable to figure out how to do that.

Reply:
I think I've convinced myself that this timeout belongs in pgwire and not in SessionClient. It was a useful exercise to understand why though!